jdaviz/app.py

```diff
@@ -254,12 +259,24 @@ def get_data_from_viewer(self, viewer_reference, data_label=None, cls=None):
         viewer = self.get_viewer(viewer_reference)
         cls = cls or viewer.default_class
 
-        data = [layer_state.layer.get_object(cls=cls)
-                if cls is not None else layer_state.layer
-                for layer_state in viewer.state.layers
-                if hasattr(layer_state, 'layer') and
-                isinstance(layer_state.layer, BaseData)
-                and (data_label is None or layer_state.layer.label == data_label)]
+        data = []
+
+        for layer_state in viewer.state.layers:
+            if hasattr(layer_state, 'layer') and \
+                (data_label is None or
+                 layer_state.layer.label == data_label):
+
+                # For raw data, just include the data itself
+                if isinstance(layer_state.layer, BaseData):
+                    data.append(layer_state.layer.get_object(cls=cls))
+
+                # For subsets, make sure to apply the subset mask to the
+                #  layer data first
+                elif isinstance(layer_state.layer, Subset):
+                    subset = layer_state.layer
+                    mask = subset.to_mask()
+                    layer_data = subset.data[mask].get_object(cls=cls)
```

Comment:
Looks like this needs to be changed to avoid a `TypeError: unhashable type: 'numpy.ndarray'`. If I change this to
```
                    component = subset.data.components[-1]
                    layer_data = subset.data[component][mask].get_object(cls=cls)
```
it resolves that error, but throws a new error since it's left trying to call get_object on an ndarray instead of a glue Data object. So right now I'm stuck with either glue erroring on trying to subset a Data object with an ndarray, or ndarray erroring due to not being a Data object. Hopefully this is easily solvable by someone with more glue familiarity...

Reply:
Spoke with @eteq last night and he thought that `get_object()` returning the original data rather than the masked data for subsets might be a Glue bug. @nmearl  @astrofrog do you have an opinion on whether this is working as intended or if should I open a Glue issue to fix the returned data from subsets?

Reply:
Sorry, why would that be a bug? Subsets do two things: they hold the original data and they hold the definition of the subset. If you retrieve the data from a subset, it is the original data. You have to apply the subset mask to get a new modified data object.

Reply:
It would be a bug if that weren't the way Subsets were intended to work. If it is that's fine, but I wanted to double check with you and @astrofrog  since @eteq thought he remembered the intended functionality differently, and since I find it counter intuitive that one has to do "extra" work to get the subsetted data from something called a Subset. I could of course have remembered that you talked about this in a previous comment...

Regardless of the answer to that question, my first comment about not being sure how to resolve the Catch-22 I encountered with the `get_object` call stands.

Reply:
``subset`` is the subset, and ``subset.data`` is a full reference to the original data - so accessing ``subset.data.get_object`` will just get a high-level object with the original data. You probably want ``subset.data.get_subset_object``? For ease of use we should probably just implement ``subset.get_object`` which would be equivalent to ``subset.data.get_subset_object``?

Reply:
Issue opened in https://github.com/glue-viz/glue/issues/2135